pandas/io/json/_normalize.py

```diff
@@ -520,7 +522,8 @@ def _recursive_extract(data, path, seen_meta, level=0):
     result = DataFrame(records)
 
     if record_prefix is not None:
-        result = result.rename(columns=lambda x: f"{record_prefix}{x}")
+        rename_fun: Callable[[Hashable], str] = lambda x: f"{record_prefix}{x}"
```

Comment:
mypy wasn't able to infer this type

Reply:
Users could presumably get the same issue when type checking their code?

what is the error if the variable type annotation is not used and what is the error if we don't change the code? (does bidirectional type inference not work here?) are lambdas tested in the MS stubs?

Reply:
> Users could presumably get the same issue when type checking their code?

Yes.  See below.
> 
> what is the error if the variable type annotation is not used and what is the error if we don't change the code? (does bidirectional type inference not work here?) 

Bidirectional type inference is subtle.  See below.

> are lambdas tested in the MS stubs?
Yes and no.

There is not a test for a `lambda` with `rename` (and I've just been copying what's being done here in the stubs over to there in my working copy),  but there are other `lambda` tests for other parts of the API.  However, the issue is quite subtle with respect to the bidirectional aspect, better illustrated with a `Dict`.    See https://github.com/python/mypy/issues/12464 .  It is bidirectional if you are creating the type inline as a parameter.  But if you store it in a variable (as shown in `rename_fun` above), you have to declare the type.

Having said that, the following code will fail with `mypy` and `pyright` with the proposed stubs here, so I think `Renamer` should just use `Callable` without specifying the types of the arguments.


```python
    df = pd.DataFrame(columns=["a"])
    df.rename(columns=lambda s: s.upper())

```

Reply:
> It is bidirectional if you are creating the type inline as a parameter. But if you store it in a variable (as shown in `rename_fun` above), you have to declare the type.

so I guess my point is, if we are needing to declare the type, why are we changing the code?

Reply:
> so I think `Renamer` should just use `Callable` without specifying the types of the arguments.

As the focus of typing has now shifted towards making the types public and typing the public API, we do need to make sure that we do not generate too many false positives.

So while in the past we tried to be as precise as possible with the type annotations, i'm fine with less precise types in these situations (for now!)

Reply:
no longer an issue with Any